### ISSUE TITLE
clang incorrectly detects a circular template recursion where there is none

### ISSUE THREAD

**Issue by llvmbot** [llvm/llvm-project#23928]
|  |  |
| --- | --- |
| Bugzilla Link | [23554](https://llvm.org/bz23554) |
| Version | trunk |
| OS | Windows NT |
| Attachments | [File containing code causing the problem](https://user-images.githubusercontent.com/60944935/143751440-3a984e84-a67f-47c8-9a3f-ef209561f051.gz) |
| Reporter | LLVM Bugzilla Contributor |
| CC | @DougGregor,@hfinkel,@zygoloid |

## Extended Description 

In the attached code clang seems to to be trying to resolve a template recursion where there is actually none in line 12 (the return type declaration using decltype). With MSVC12 (2013) and gcc4.9 the same code compiles without any errors or warnings.

Workaround: Replacing the explicit decltype declaration with C++14's decltype(auto) results in compilation without errors.

I tested this with clang3.6 on OSX and the 3.7 trunk version on Windows in MSVC emulation mode with the same outcome.

**Comment by wheatman** [comment]
This was fixed between 3.9.1 and 4.0.0
https://godbolt.org/z/5E845beT4
code
```
template<typename WrapperT, typename WrappedT>
struct Proxy {
    virtual WrappedT& valueRef() = 0;
    virtual const WrappedT& valueCRef() const = 0;
    virtual ~Proxy() = default;
};
 
template<typename WrapperT, typename WrappedT>
struct Add {
    template<typename RhsT>
    friend auto operator + (const Proxy<WrapperT, WrappedT>& lhs, const RhsT& rhs)
        -> decltype(lhs.valueCRef() + rhs) {
 
        return lhs.valueCRef() + rhs;
    }
};
 
template<typename T>
struct Wrapper 
    : Proxy<Wrapper<T>, T>
    , Add<Wrapper<T>, T> {
 
    T& valueRef() override { return val; }
    const T& valueCRef() const override { return val; }
 
    operator T& () { return valueRef(); }
    operator const T& () const { return valueCRef(); } 
private:
    T val;
};
 
int main() {
    Wrapper<int> a;
    (void)(a + a);
}

```

old error
```
<source>:12:37: fatal error: recursive template instantiation exceeded maximum depth of 256
        -> decltype(lhs.valueCRef() + rhs) {
                                    ^
<source>:12:37: note: while substituting deduced template arguments into function template 'operator+' [with RhsT = Wrapper<int>]
<source>:12:37: note: while substituting deduced template arguments into function template 'operator+' [with RhsT = Wrapper<int>]
<source>:12:37: note: while substituting deduced template arguments into function template 'operator+' [with RhsT = Wrapper<int>]
<source>:12:37: note: while substituting deduced template arguments into function template 'operator+' [with RhsT = Wrapper<int>]
<source>:12:37: note: while substituting deduced template arguments into function template 'operator+' [with RhsT = Wrapper<int>]
<source>:12:37: note: (skipping 247 contexts in backtrace; use -ftemplate-backtrace-limit=0 to see all)
<source>:12:37: note: while substituting deduced template arguments into function template 'operator+' [with RhsT = Wrapper<int>]
<source>:12:37: note: while substituting deduced template arguments into function template 'operator+' [with RhsT = Wrapper<int>]
<source>:12:37: note: while substituting deduced template arguments into function template 'operator+' [with RhsT = Wrapper<int>]
<source>:12:37: note: while substituting deduced template arguments into function template 'operator+' [with RhsT = Wrapper<int>]
<source>:34:14: note: while substituting deduced template arguments into function template 'operator+' [with RhsT = Wrapper<int>]
    (void)(a + a);
             ^
1 error generated.
Compiler returned: 1
```

**Comment by llvmbot** [comment]
@llvm/issue-subscribers-clang-frontend

Author: None (llvmbot)

<details>
|  |  |
| --- | --- |
| Bugzilla Link | [23554](https://llvm.org/bz23554) |
| Version | trunk |
| OS | Windows NT |
| Attachments | [File containing code causing the problem](https://user-images.githubusercontent.com/60944935/143751440-3a984e84-a67f-47c8-9a3f-ef209561f051.gz) |
| Reporter | LLVM Bugzilla Contributor |
| CC | @DougGregor,@hfinkel,@zygoloid |

## Extended Description 


In the attached code clang seems to to be trying to resolve a template recursion where there is actually none in line 12 (the return type declaration using decltype). With MSVC12 (2013) and gcc4.9 the same code compiles without any errors or warnings.

Workaround: Replacing the explicit decltype declaration with C++14's decltype(auto) results in compilation without errors.

I tested this with clang3.6 on OSX and the 3.7 trunk version on Windows in MSVC emulation mode with the same outcome.
</details>
